porcelain/miner.go

```diff
@@ -12,13 +12,146 @@ import (
 	cbor "gx/ipfs/QmcZLyosDwMKdB6NLRsiss9HXzDPhVhhRtPy67JFKTDQDX/go-ipld-cbor"
 
 	minerActor "github.com/filecoin-project/go-filecoin/actor/builtin/miner"
+	"github.com/filecoin-project/go-filecoin/actor/builtin/storagemarket"
 	"github.com/filecoin-project/go-filecoin/address"
 	"github.com/filecoin-project/go-filecoin/exec"
 	"github.com/filecoin-project/go-filecoin/types"
 	vmErrors "github.com/filecoin-project/go-filecoin/vm/errors"
 	w "github.com/filecoin-project/go-filecoin/wallet"
 )
 
+// mcAPI is the subset of the plumbing.API that MinerCreate uses.
+type mcAPI interface {
+	ConfigGet(dottedPath string) (interface{}, error)
+	ConfigSet(dottedPath string, paramJSON string) error
+	GetAndMaybeSetDefaultSenderAddress() (address.Address, error)
+	MessageQuery(ctx context.Context, optFrom, to address.Address, method string, params ...interface{}) ([][]byte, *exec.FunctionSignature, error)
+	MessageSend(ctx context.Context, from, to address.Address, value *types.AttoFIL, gasPrice types.AttoFIL, gasLimit types.GasUnits, method string, params ...interface{}) (cid.Cid, error)
+	MessageWait(ctx context.Context, msgCid cid.Cid, cb func(*types.Block, *types.SignedMessage, *types.MessageReceipt) error) error
+	NetworkGetPeerID() peer.ID
+	WalletFind(address address.Address) (w.Backend, error)
+	WalletGetPubKeyForAddress(addr address.Address) ([]byte, error)
+	WalletHasAddress(addr address.Address) bool
+}
+
+// MinerCreate creates a new miner actor for the given account and returns its address.
+// It will wait for the the actor to appear on-chain and add set the address to mining.minerAddress in the config.
+// TODO: add ability to pass in a KeyInfo to store for signing blocks.
+//       See https://github.com/filecoin-project/go-filecoin/issues/1843
+func MinerCreate(
+	ctx context.Context,
+	plumbing mcAPI,
+	accountAddr address.Address,
+	gasPrice types.AttoFIL,
+	gasLimit types.GasUnits,
+	pledge uint64,
+	pid peer.ID,
+	collateral *types.AttoFIL,
+) (_ *address.Address, err error) {
+	if accountAddr == (address.Address{}) {
+		accountAddr, err = plumbing.GetAndMaybeSetDefaultSenderAddress()
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if pid == "" {
+		pid = plumbing.NetworkGetPeerID()
+	}
+
+	ctx = log.Start(ctx, "Node.CreateMiner")
+	defer func() {
+		log.FinishWithErr(ctx, err)
+	}()
+
+	addrInterface, err := plumbing.ConfigGet("mining.minerAddress")
+	if err != nil {
+		return nil, err
+	}
+	addr, success := addrInterface.(address.Address)
+	if !success {
+		return nil, fmt.Errorf("failed to read miner address")
+	}
+	if (addr != address.Address{}) {
+		return nil, fmt.Errorf("can only have one miner per node")
+	}
+
+	pubKey := []byte{}
+	if plumbing.WalletHasAddress(addr) {
```

Comment:
BLOCKING: This should be the minerOwner address and it should be an error if the public key doesn't exist.